### PR TITLE
fix: verify done_blob_uri consistency on RawImpressionUpload create idempotency

### PR DESCRIPTION
**Pull request by getina**
On a `request_id` idempotency match, `createRawImpressionUpload` returned the existing row without verifying field consistency; per AIP-155 it now throws `ALREADY_EXISTS` when the existing `done_blob_uri` differs from the request. This prevents a reused `request_id` with a different `done_blob_uri` from returning the wrong upload and a TEE labeling the wrong files.

Issue: #4117

### CODE REVIEW

**Comment by wfa-reviewable** [comment]
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4118)

File: src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt [hunk 1]
```diff
@@ -128,6 +128,8 @@ class RawImpressionUploadService(
           InternalErrors.Reason.RANKER_JOB_STATE_INVALID,
           InternalErrors.Reason.RANK_INDEX_BLOB_NOT_FOUND,
           InternalErrors.Reason.RANK_INDEX_BLOB_ALREADY_EXISTS,
+          InternalErrors.Reason.RAW_IMPRESSION_UPLOAD_ALREADY_EXISTS ->
```
Comment (stevenwarejones) [inline src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt:138]: blocking: the diff broke the `when` structure. The trailing comma after `RANK_INDEX_BLOB_ALREADY_EXISTS,` joins it (and the 28 reasons above it) with the new `RAW_IMPRESSION_UPLOAD_ALREADY_EXISTS ->` arm, so every reason listed now maps to `Status.ALREADY_EXISTS` instead of falling through to `null -> Status.INTERNAL`.

Concretely, post-PR these all return `ALREADY_EXISTS` instead of `INTERNAL`:

- `DATA_PROVIDER_MISMATCH`
- `ETAG_MISMATCH` (was `INTERNAL` before — now ALREADY_EXISTS)
- `REQUIRED_FIELD_NOT_SET`
- `INVALID_FIELD_VALUE`
- `IMPRESSION_METADATA_NOT_FOUND`, `RAW_IMPRESSION_UPLOAD_NOT_FOUND`, etc.
- ~24 more

This breaks the `VidLabelingDispatcher`: `createRawImpressionUpload` catches `ALREADY_EXISTS` and tries to recover by `findUploadByDoneBlobUri(doneBlobPath)` — a genuinely transient internal error from the service now triggers the recovery path, silently masking the failure.

Split the new arm into its own branch with its own body:

```kotlin
throw when (InternalErrors.getReason(e)) {
  InternalErrors.Reason.RAW_IMPRESSION_UPLOAD_ALREADY_EXISTS ->
    Status.ALREADY_EXISTS.withCause(e).asRuntimeException()
  InternalErrors.Reason.DATA_PROVIDER_MISMATCH,
  InternalErrors.Reason.IMPRESSION_METADATA_NOT_FOUND,
  // ... rest of the prior list ...
  InternalErrors.Reason.RANK_INDEX_BLOB_ALREADY_EXISTS,
  null -> Status.INTERNAL.withCause(e).asRuntimeException()
}
```

Add a test that asserts a `null`-reason internal failure surfaces as `INTERNAL` from this RPC, so this trap is caught next time.

File: src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt [hunk 2]
```diff
@@ -189,6 +191,8 @@ class RawImpressionUploadService(
...
```
Comment (stevenwarejones) [inline src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt]: blocking: same `when`-fallthrough bug as the inline on line 131 — every reason listed above `RAW_IMPRESSION_UPLOAD_ALREADY_EXISTS ->` now maps to `Status.ALREADY_EXISTS` instead of `Status.INTERNAL`.

Also: this is `getRawImpressionUpload` (a read RPC). It never produces `RAW_IMPRESSION_UPLOAD_ALREADY_EXISTS` internally — adding the reason here is dead code at best, and the structural break makes it actively wrong. Drop the new reason from this site entirely; the `null ->` fallthrough already maps unknown reasons to `INTERNAL`.

File: src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt [hunk 3]
```diff
@@ -296,6 +300,8 @@ class RawImpressionUploadService(
...
```
Comment (stevenwarejones) [inline src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt]: blocking: same `when`-fallthrough bug as the inline on line 131. Same reasoning as the inline on line 194: `listRawImpressionUploads` is a read RPC and never produces `RAW_IMPRESSION_UPLOAD_ALREADY_EXISTS`. Drop the new reason from this site.

File: src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt (hunk 1 above)
Comment (getina) [reply src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt:138]: done - thanks for catching

File: src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt (hunk 2 above)
Comment (getina) [reply src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt]: when I remove RAW_IMPRESSION_UPLOAD_ALREADY_EXISTS, i get error: 'when' expression must be exhaustive. this reason need to be here unless you prefer using else

File: src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt (hunk 3 above)
Comment (getina) [reply src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt]: when I remove RAW_IMPRESSION_UPLOAD_ALREADY_EXISTS, i get error: 'when' expression must be exhaustive. this reason need to be here unless you prefer using else

**Review by getina — COMMENTED** [review]
> @getina made 3 comments.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4118)* status: 0 of 13 files reviewed, 3 unresolved discussions (waiting on stevenwarejones).

**Review by stevenwarejones — COMMENTED** [review]
> @stevenwarejones reviewed 13 files and all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4118)* status: all files reviewed, 3 unresolved discussions (waiting on getina).

File: src/main/kotlin/org/wfanet/measurement/edpaggregator/service/internal/Errors.kt [hunk 4]
```diff
@@ -292,6 +293,17 @@ class RawImpressionUploadNotFoundException(
...
+class RawImpressionUploadAlreadyExistsException(
+  dataProviderResourceId: String,
+  cause: Throwable? = null,
+) :
+  ServiceException(
+    Errors.Reason.RAW_IMPRESSION_UPLOAD_ALREADY_EXISTS,
+    "RawImpressionUpload already exists for DataProvider $dataProviderResourceId",
+    mapOf(Errors.Metadata.DATA_PROVIDER_RESOURCE_ID to dataProviderResourceId),
+    cause,
+  )
```
Comment (stevenwarejones) [inline src/main/kotlin/org/wfanet/measurement/edpaggregator/service/internal/Errors.kt:396]: nit: when this fires, an operator sees `"RawImpressionUpload already exists for DataProvider edp1"` and has nothing to diagnose with — they have to query Spanner to find which row collided and what the existing `done_blob_uri` is vs the request's. The two pieces of context that turn this into an actionable error are:

- the **`request_id`** that collided (so they can trace back to `RequestIds.forRawImpressionUpload(doneBlobPath, generation)` and find the deterministic-UUID collision source)
- both **`done_blob_uri`** values (existing row's vs incoming request's) so the divergence is visible without a Spanner query

Something like:

```kotlin
class RawImpressionUploadAlreadyExistsException(
  dataProviderResourceId: String,
  requestId: String,
  existingDoneBlobUri: String,
  incomingDoneBlobUri: String,
  cause: Throwable? = null,
) :
  ServiceException(
    Errors.Reason.RAW_IMPRESSION_UPLOAD_ALREADY_EXISTS,
    "RawImpressionUpload already exists for DataProvider $dataProviderResourceId with " +
      "request_id $requestId: existing done_blob_uri=$existingDoneBlobUri, " +
      "incoming done_blob_uri=$incomingDoneBlobUri",
    mapOf(
      Errors.Metadata.DATA_PROVIDER_RESOURCE_ID to dataProviderResourceId,
      Errors.Metadata.CREATE_REQUEST_ID to requestId,
    ),
    cause,
  )
```

Needs a new `CREATE_REQUEST_ID` metadata key. Same context will be useful when the pattern propagates to `createPoolAssignmentJob` (#4044) and other Create RPCs.

Not blocking.

File: src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt (hunk 1 above)
Comment (stevenwarejones) [reply src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt:138]: Verified at head c2dc9ed7. The new arm is now split into its own body branch (line 131), so all the prior reasons fall through to `null -> Status.INTERNAL` and only `RAW_IMPRESSION_UPLOAD_ALREADY_EXISTS` maps to `Status.ALREADY_EXISTS`. The added regression test (`createRawImpressionUpload surfaces INTERNAL when the internal failure has no reason`) locks in the correct semantics. Fixed.

File: src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt (hunk 2 above)
Comment (stevenwarejones) [reply src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt]: Verified at head c2dc9ed7. You're right that dropping the reason breaks Kotlin's `when`-exhaustiveness — the correct fix is exactly what you did: list `RAW_IMPRESSION_UPLOAD_ALREADY_EXISTS` under the `null -> Status.INTERNAL` arm here. Get is a read RPC that doesn't produce this reason, so mapping it to INTERNAL is correct fallthrough. My suggestion to "drop the reason" was wrong. Fixed.

File: src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt (hunk 3 above)
Comment (stevenwarejones) [reply src/main/kotlin/org/wfanet/measurement/edpaggregator/service/v1alpha/RawImpressionUploadService.kt]: Verified at head c2dc9ed7. Same correct fix as gh-3501049190 — `RAW_IMPRESSION_UPLOAD_ALREADY_EXISTS` listed under `null -> Status.INTERNAL` so the read RPC's exhaustive `when` compiles and any actual occurrence (which shouldn't happen on List) surfaces as INTERNAL. Fixed.

**Review by stevenwarejones — APPROVED** [review]

File: src/main/kotlin/org/wfanet/measurement/edpaggregator/service/internal/Errors.kt (hunk 4 above)
Comment (getina) [reply src/main/kotlin/org/wfanet/measurement/edpaggregator/service/internal/Errors.kt:396]: done

**Review by getina — COMMENTED** [review]
> @getina made 1 comment and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4118)* status: all files reviewed, 3 unresolved discussions (waiting on getina).

**Review by stevenwarejones — APPROVED** [review]

File: src/main/kotlin/org/wfanet/measurement/edpaggregator/service/internal/Errors.kt (hunk 4 above)
Comment (stevenwarejones) [reply src/main/kotlin/org/wfanet/measurement/edpaggregator/service/internal/Errors.kt:396]: Verified at head 6012c54f — RawImpressionUploadAlreadyExistsException now carries requestId + existingDoneBlobUri + incomingDoneBlobUri (Errors.kt:298-313), and CREATE_REQUEST_ID metadata key added at line 82. Error message now names the collision source and both divergent URIs. Fixed.

**Review by stevenwarejones — COMMENTED** [review]
> @stevenwarejones+AGNT made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4118)* status: 10 of 13 files reviewed, 3 unresolved discussions (waiting on getina, SanjayVas, and stevenwarejones).

**Review by stevenwarejones — APPROVED** [review]
> @stevenwarejones reviewed 3 files and all commit messages, and resolved 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4118)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on SanjayVas).

**Review by SanjayVas — APPROVED** [review]
> @SanjayVas reviewed 13 files and all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4118)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on getina).

**Review by getina — COMMENTED** [review]
> @getina reviewed 15 files and all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4118)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on getina).
